onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+    ORT_ENFORCE(!(graph.IsInputsIncludingInitializers(node_arg) && graph.IsOutput(node_arg)));
+    NodeArg& cast_input = !is_node_arg_cast_output ? *node_arg : new_node_arg;
+    NodeArg& cast_output = is_node_arg_cast_output ? *node_arg : new_node_arg;
+    const std::vector<NodeArg*> cast_inputs = {&cast_input};
+    const std::vector<NodeArg*> cast_outputs = {&cast_output};
+    ONNX_NAMESPACE::AttributeProto to_attribute;
+    to_attribute.set_name("to");
+    to_attribute.set_type(ONNX_NAMESPACE::AttributeProto::INT);
+    to_attribute.set_i(static_cast<int64_t>(data_type));
+    NodeAttributes attributes({{"to", to_attribute}});
+
+    Node& cast = graph.AddNode(graph.GenerateNodeName(node_arg->Name() + "_cast"),
+                               "Cast",
+                               "Created a new Cast node",
+                               cast_inputs,
+                               cast_outputs,
+                               &attributes);
+    inserted_node_names.push_back(cast.Name());
+    Node* producer = graph.GetMutableProducerNode(node_arg->Name());
+    std::vector<Node*> consumers = graph.GetMutableConsumerNodes(node_arg->Name());
+    int output_index = (producer != nullptr) ? optimizer_utils::IndexOfNodeOutput(*producer, *node_arg) : -1;
+    // Update consumers of node_arg to use the output of the cast node
+    int cast_output_index = optimizer_utils::IndexOfNodeOutput(cast, cast_output);
+    for (Node* consumer : graph.GetMutableConsumerNodes(node_arg->Name())) {
+      if (consumer != nullptr &&
+          std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end()) {
+        auto& consumer_inputs = consumer->MutableInputDefs();
+        int input_index = optimizer_utils::IndexOfNodeInput(*consumer, *node_arg);
+        if (producer != nullptr) {
+          graph.RemoveEdge(producer->Index(), consumer->Index(), output_index, input_index);
+        }
+        std::replace(consumer_inputs.begin(), consumer_inputs.end(), &cast_input, &cast_output);
+        graph.AddEdge(cast.Index(), consumer->Index(), cast_output_index, input_index);
+      }
+    }
+    if (producer != nullptr) {
+      auto& producer_outputs = producer->MutableOutputDefs();
+      std::replace(producer_outputs.begin(), producer_outputs.end(), &cast_output, &cast_input);
+      graph.UpdateProducerNode(cast_input.Name(), producer->Index());
+      int input_index = optimizer_utils::IndexOfNodeInput(cast, cast_input);
+      graph.AddEdge(producer->Index(), cast.Index(), output_index, input_index);
+    }
+    graph.UpdateProducerNode(cast_output.Name(), cast.Index());
+  }
+  return Status::OK();
+}
+
+// RemoveCastNodes
+// Remove the cast nodes specified in casts vector and fix the graph edges accordingly.
+static Status RemoveCastNodes(Graph& graph, std::vector<Node*> casts, std::deque<onnxruntime::NodeIndex>& removed_nodes) {
+  ORT_ENFORCE(casts.size() > 0);
```

Comment:
there is acctually a hidden condition to use this function, all the casts node has to come in a sequence. 
Let's mark this in the comment to avoid misuse... 

Maybe rename this function to reflect reality, RemoveSequentialCastNodes ?

Reply:
Renaming to RemoveCastNodeChain